### PR TITLE
[models] Add django swapper #3

### LINKED ISSUE #3: [firmware-upgrader] Abstract and Swappable models
- Extract the base logic of the models into abstract models
- Set up django swappable models like we do in django-ipam / django-freeradius
- Set up a sample_app for testing custom apps extending the models, like we do in django-ipam / openwisp-ipam

### PR DESCRIPTION
**Pull request by atb00ker**
Use swapper to make it `openwisp-firmware-upgrader` a reusable application.

Closes #3

### CODE REVIEW

**Comment by atb00ker** [comment]
> The `sample_app` will confirm if there's any issue that needs to be fixed.

Done! :smile:

This was a great issue, I learned that swapper can be used in the `migrations` and hence I **edited the existing migrations** since the app is not yet released! :smile:

File: runtests.py [hunk 1]
```diff
@@ -11,5 +11,8 @@
     from django.core.management import execute_from_command_line
     args = sys.argv
     args.insert(1, "test")
-    args.insert(2, "openwisp_firmware_upgrader")
+    if not os.environ.get('SAMPLE_APP', False):
+        args.insert(2, "openwisp_firmware_upgrader")
+    else:
+        args.insert(2, "openwisp2")
```
[the remark this answers is not in the document]
Comment (nemesisdesign) [reply runtests.py]: Let me correct this.. I just noticed that's the project.
We need a new app in that directory named sample_app, like we have in django_ipam: https://github.com/openwisp/django-ipam/tree/master/tests/sample_ipam

Ok, there we have `sample_ipam`, maybe we can have `sample_firmware_upgrader`?

Also can you add a mention of this thing in the test section of the README?
That is, explaining very briefly what this app is for (needed to test the extensibility).

[the remark this answers is not in the document]
Comment (atb00ker) [reply runtests.py]: My understand is that you mean adding an app inside the existing project `openwisp2` and not renaming that project. If no the other hand you meant renaming the project, please let me know, I've that in a different branch as well! :smile:

**Comment by atb00ker** [comment]
> BTW. no README updates regarding the changes you're introducing? That'd be a shame because not many would take advantage of these.

Oh, of course let me do that, thanks for reminding! :smile:

File: README.rst [hunk 2]
```diff
@@ -223,6 +223,98 @@ Run tests with:
...
+    pip install openwisp-utils
+
+2. Add ``EXTENDED_APPS``
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Add the following to your ``settings.py``:
+
+.. code-block:: python
+
+    EXTENDED_APPS = ('openwisp-firmware-upgrader',)
```
Comment (nemesisdesign) [inline README.rst]: this doesn't look right, I think it should be as: https://github.com/openwisp/openwisp-firmware-upgrader/pull/9/files#diff-9dec27df1ed1039b50d2b96520f1de08R199

**Comment by nemesisdesign** [comment]
I did the changes myself and merged it. Please take a look at the minor changes for future reference: https://github.com/openwisp/openwisp-firmware-upgrader/commit/938078243258ccbed7fdedd9d4e120204876117c

File: README.rst (hunk 2 above)
Comment (atb00ker) [reply README.rst]: Yeah, my bad! :smile:
Thanks for fixing it! :smile:

**Pull request merged**